File: scr/Solution005.java
package p5;
//https://gist.github.com/bnyu/a79f4548acfcaa99d4a3ac6b3f535e27
// Accepted

/**
 * 5. Longest Palindromic Substring
 * Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.
 */
class Solution {
    public String longestPalindrome(String s) {
        String ans;
        if (!s.isEmpty()) {
            char[] chars = s.toCharArray();
            int length = chars.length;
            int longest = 1;
            ans = String.valueOf(chars[0]);
            for (int i = 1; i < length; i++) {
                int len = 1;
                for (int j = 1; i - j >= 0 && i + j < length; j++) {
                    if (chars[i - j] != chars[i + j])
                        break;
                    len += 2;
                }
                if (len > longest) {
                    ans = String.valueOf(chars, i - len / 2, len);
                    longest = len;
                }

                len = 0;
                for (int j = 1; i - j >= 0 && i + j - 1 < length; j++) {
                    if (chars[i - j] != chars[i + j - 1])
                        break;
                    len += 2;
                }
                if (len > longest) {
                    ans = String.valueOf(chars, i - len / 2, len);
                    longest = len;
                }
            }
        } else
            ans = "";
        return ans;
    }
}

File: scr/Solution005.kt
/**
 * 5. Longest Palindromic Substring
 * Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.
 */
class Solution005 {
    fun longestPalindrome(s: String): String {
        val ans: String
        if (!s.isEmpty()) {
            val chars = s.toCharArray()
            val length = chars.size
            var longest = 1
            var leftIndex = 0
            var rightIndex = 0
            for (i in 1 until length) {
                //以第i个为中心 往左往右依次对比
                var len = 1
                var j = 1
                while (i - j >= 0 && i + j < length) {
                    if (chars[i - j] != chars[i + j])
                        break
                    len += 2
                    j++
                }
                if (len > longest) {
                    leftIndex = i - len / 2
                    rightIndex = len
                    longest = len
                }
                //以第i个与第i-1个之间为中心
                len = 0
                j = 1
                while (i - j >= 0 && i + j - 1 < length) {
                    if (chars[i - j] != chars[i + j - 1])
                        break
                    len += 2
                    j++
                }
                if (len > longest) {
                    leftIndex = i - len / 2
                    rightIndex = len
                    longest = len
                }
            }
            ans = if (longest != 1) String(chars, leftIndex, rightIndex) else chars[0].toString()
        } else
            ans = ""
        return ans
    }
}

